Update to API/Routes Design Pattern, Adding Login/Signup API, Tank API **Be sure to title your PRs next time :)**

Airor4 opened the pull request:
**Description**
Added proxy, and changed the api/routes to a route/controller design pattern. User login and signup are ready to go. See the route files on how to call them. When testing the API, make sure you are running the backend server along with the frontend server. It may seem obvious but we can be forgetful hahah.

Fixes #(issue)
#43 #44 

**Type of change**
Check options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Fix or feature that would cause existing functionality to not work as expected
- [ ] This change requires a update in the design document

**Checklist:**
Go down the list and check them off.

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the design document (if applicable)
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

Commit:
added json webtoken to signup and checking for valid signup parameters

Commit:
Added jwt authentication and login api

Commit:
added flow to necessary js files

Commit:
Added flow, made null checks, and added documentation

Commit:
conforms to flow standards

Commit:
pull from remote Baylor

Commit:
whole lotta changes

Commit:
ALL WORKING

Commit:
json secret placeholder - needs to be set up securely at production

Commit:
started out on controllers and trying to connect api

Commit:
Bring in Jorge's login with api config

Commit:
Added proxy, Front-end: 3000 back-end 3001. Added controllers

Commit:
.gitignore fixed to not track package-lock

Commit:
pull from master so that i have emil's updated models

Commit:
updated package.json

Commit:
babel configuration

Commit:
added regen-runtime to get rid of error and fixed front end api path

Commit:
Proxy completed and switched to controller/route design pattern

Commit:
Added proxy and changed design pattern 2

Commit:
got rid of api subdirectory

Commit:
merge from master, removed jorge's code from login.js

Commit:
moved up a directory

Commit:
Moved up to Fortuna

Commit:
updated tank controller

Commit:
recovered updates to code

Commit:
pull from master??

Commit:
merge from origin/master

Commit:
.gitignore fix for /lib/

Commit:
added files back

Commit:
Deleted useless package and conformed to flow

SecondThread commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,27 +0,0 @@
----
-name: Bug report
```
I'm pretty sure we don't want this deleted. @SleepyLime to confirm though

Airor4 replied:
I'm not sure how this was deleted or how to get it back.

SecondThread commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -1,17 +0,0 @@
----
-name: Feature request
```
Same thing; pretty sure we want this to still be here. @SleepyLime

SecondThread commented on .github/pull_request_template.md:
```diff
@@ -1,22 +0,0 @@
-**Description**
-Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
```
Ditto

SecondThread commented on src/backend/controllers/tankController.js:
```diff
@@ -0,0 +1,49 @@
+  } from 'express';
+
+exports.getFavorite = (req: $Request, res: $Response) => {
+    user.findById(req.params.userId, 'favoriteTankId', function(err, myUser){
```
What are the types of err and myUser? Can you flow-type them please?

SecondThread commented on src/backend/controllers/userController.js:
```diff
@@ -0,0 +1,169 @@
+    const { userName, password } = req.body;
+
+    try{
+        // See if User exists - might change this to const
```
Yeah, change this to const please :)

SecondThread commented on src/backend/controllers/userController.js:
```diff
@@ -0,0 +1,169 @@
+        }
+}
+
+/*exports.updateCassus = async (req: $Request, res: $Response){
```
Delete this if it isn't working and then publish it in a later push if you need to

SecondThread commented on src/backend/controllers/userController.js:
```diff
@@ -0,0 +1,169 @@
+
+exports.checkAuth = async (req: $Request, res: $Response) => {
+    try{
+        const user = await User.findById(req.userId).select('-password');
```
Just curious: what is the hyphen for before password?

Airor4 replied:
It excludes the password document field.

SecondThread replied:
oh lol that's probably important

SecondThread commented on src/backend/routes/userRoutes.js:
```diff
@@ -0,0 +1,55 @@
+// Req must include userName and password in body
+// Returns a jwt upon success
+router.post('/registerUser', [
+    check('userName', 'Please enter a username with 3 or more characters')
```
Clean, I like it

SecondThread reviewed (changes requested):
I left comments for some of the things I think you should change.

Also, did you test this stuff to make sure it works? Can you write a doc somewhere so we can see what we need to do to verify that your stuff works?

Derplime commented:
@Airor4 

So for recovering the `.github` file, you could see if it exists in another branch (i.e. `master`), copy it locally to your computer, go back to your branch, paste that in, and push that up to your remote. Assuming you didn't change any files in that `.github` file, that should remove that change compared to master.

Derplime commented on src/.env:
```diff
@@ -0,0 +1,2 @@
+SECRET=supahsecret4
+PORT=3001
```
Request new line at end of file.

Derplime commented on src/login/Login.css:
```diff
@@ -0,0 +1,90 @@
+.infopopup {
+    padding: 5px 50px 20px 50px;
+    background: #004e92;
+}
```
Another new line here please. Sorry for pickiness.

Derplime reviewed (changes requested):
So outside the changes @SecondThread requested, and my stupid picky ones, I'd also second on having a set of instructions to test this correctly before giving my approval.

Commit:
added flow type annotations on callbacks

Commit:
Got back .github

Airor4 committed:
Merge branch 'master' into Baylor

sleepylime committed:
Update userRoutes.js

changed Retrieve to retrieve, removed boilerplate code for retrieveAll and update

Airor4 committed:
Update tankController.js

updated function for adding favoriteTank. works correctly now

Derplime committed:
fixed mongoose depreciation warning

SecondThread reviewed (approved):
lgtm

Derplime reviewed (approved):
👍  Forgot to put mine in as well.

Derplime commented:
Merging this myself since I'll need this to continue my work.
